tests/ha_test.go

```diff
@@ -0,0 +1,147 @@
+}
+
+func (suite *HighAvailabilityTestSuite) TestParameters() {
+	operator := suite.operator
+	params := suite.leaderElectionParams
+	args, err := utils.Kubectl("get", "deployment", operator.InstanceName, "-n", operator.Namespace,
+		"-o=jsonpath={.spec.template.spec.containers[0].args}")
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+	availableReplicas, _ := utils.Kubectl("get", "deployment", operator.InstanceName, "-n", operator.Namespace,
+		"-o=jsonpath={.status.availableReplicas}")
+
+	suite.Equal(params["replicas"], availableReplicas)
+	suite.Contains(args, fmt.Sprint("-leader-election=", params["enableLeaderElection"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-lock-name=", params["leaderElectionLockName"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-lock-namespace=", operator.Namespace))
+	suite.Contains(args, fmt.Sprint("-leader-election-lease-duration=", params["leaderElectionLeaseDuration"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-renew-deadline=", params["leaderElectionRenewDeadline"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-retry-period=", params["leaderElectionRetryPeriod"]))
+}
+
+func (suite *HighAvailabilityTestSuite) TestLeaderElection() {
+	leaderElectionRecord, err := getLeaderElectionRecord(suite.operator)
+	if suite.NoError(err) {
+		suite.NotEmpty(leaderElectionRecord.HolderIdentity)
+	}
+}
+
+func (suite *HighAvailabilityTestSuite) TestFailover() {
+	operator := suite.operator
+	leaderElectionRecord, err := getLeaderElectionRecord(operator)
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+	fmt.Println("Current leader: ", leaderElectionRecord.HolderIdentity)
+	// deploy workload
+	jobName := "mock-task-runner"
+	mockTaskRunner := utils.SparkJob{
+		Name:     jobName,
+		Template: "spark-mock-task-runner-job.yaml",
+		Params: map[string]interface{}{
+			"args": []string{"1", "30"},
+		},
+	}
+	if err := operator.SubmitJob(&mockTaskRunner); err != nil {
+		suite.FailNow(err.Error())
+	}
+	// wait until the application is running
+	if err := operator.WaitForJobState(mockTaskRunner, v1beta2.RunningState); err != nil {
+		suite.FailNow(err.Error())
+	}
+
+	log.Infof("deleting current leader pod \"%s\"", leaderElectionRecord.HolderIdentity)
+	err = utils.DeleteResource(operator.Namespace, "pod", leaderElectionRecord.HolderIdentity)
+
+	// check re-election
+	if err := utils.RetryWithTimeout(electionRecordRetryTimeout, electionRecordRetryInterval, func() error {
+		if newLeaderElectionRecord, err := getLeaderElectionRecord(operator); err != nil {
+			return err
+		} else if newLeaderElectionRecord.HolderIdentity == leaderElectionRecord.HolderIdentity {
+			return errors.New("Waiting for the new leader to be elected")
+		} else {
+			log.Info("New leader found: ", newLeaderElectionRecord.HolderIdentity)
+		}
+		return nil
+	}); err != nil {
+		suite.FailNow(err.Error())
+	}
+
+	suite.NoError(operator.WaitForJobState(mockTaskRunner, v1beta2.CompletedState))
```

Comment:
Once Spark job is submitted in cluster mode,  technically, it doesn't need a client machine/pod to be up anymore and can complete independently. What would be nice to verify is that the original leader submit the job and the newly elected leader handles job status update when it completes. Otherwise, in the current implementation, there's no link between failover and status update handling. 

How complicated would it be to implement that testing scenario?

As an alternative, we can simplify it a bit and submit two independent jobs before and after reelection and verify they're handled by different controller instances.

Reply:
> Once Spark job is submitted in cluster mode, technically, it doesn't need a client machine/pod to be up anymore and can complete independently. 

If we talking about the driver/executor pods, yes, they complete independently and don't require Spark operator instance to perform status updates. However, `WaitForJobState` method is checking the status of `SparkApplication` resource, not a `Pod`, and the former is updated by the operator controller of a new leader. 
But I agree, that checking the logs to verify the newly elected leader handles job status update would be more explicit for this case.

Reply:
SGTM, thanks for the clarification 👍